### PR TITLE
make clamp_min/max use minimum/maximum kernels, make clamp* correctly propagate nans

### LINKED ISSUE #67428: Avoid converting Tensors to Scalars for ops that have Tensor (including optional) argument overload
## 🚀 Feature
Avoid converting Tensors to Scalars for ops that have Tensor variance

## Motivation
PyTorch has many ops that have both Scalar and Tensor variations. Sometimes, the overload is selected that converts a Tensor to Scalar, which causes a Tensor value from a backend device to be fetched onto the host for getting the Scalar argument.

This causes a bubble in execution (if we create lazy execution graphs, then these cause the graph to break as well). It would be good to allow ops to work on Tensor arguments, so that the ops can continue to work on tensors that are already in the backend device.

## Pitch

As per the order of overloads, Tensor overloads come first, before Scalar overloads. However, in case of the optional arguments like Scalar? and Tensor? variations (example: clamp), the order is not determined similarly. The optional argument ordering should also be similarly handled (#67429) 

Along with that, PyTorch also has an internal list of ops which allow converting numbers to Tensors. If more ops are allowed to convert numbers to Tensors, then it would reduce the number of cases Tensor values need to be sent to host. (#67435)

Example:
>>> import torch
>>> a = torch.Tensor(2, 3)
>>> b = torch.Tensor(3)
>>> c = a.clamp(0, b[0] - 1)  <- Here, clamp favors the Scalar? overload and gets the b[0] to host as a scalar value. Instead, it should be possible to send both the arguments as Tensors so that the execution continues to be on device.

### PR DESCRIPTION
**Pull request by ngimel**
Since clamp_min and maximum is the same op, reuse the same kernel (it also correctly propagate nans from both input and boundary, clamp* propagated from input only). 
Also fixed codegen to make Tensor? overloads come before Scalar? overloads, cc @alband. 
Fixes #67428 and #76795 (scalar overloads for clamp* are still not fixed, will do in the next PR).

### CODE REVIEW

File: aten/src/ATen/native/cpu/TensorCompareKernel.cpp [hunk 1]
```diff
@@ -335,10 +335,14 @@ static void clamp_kernel_impl(TensorIteratorBase& iter) {
...
-        return std::min(std::max(a, min), max);
+        if (min != min || max != max) {
+            return std::numeric_limits<scalar_t>::quiet_NaN();
+        } else {
+            return std::min(std::max(a, min), max);
+        }
       },
       [](Vectorized<scalar_t> a, Vectorized<scalar_t> min, Vectorized<scalar_t> max) {
-        return vec::clamp(a, min, max);
+        return vec::minimum(vec::maximum(a, min), max);
```
Comment (albanD) [inline aten/src/ATen/native/cpu/TensorCompareKernel.cpp:345]: These handle nan properly?

File: aten/src/ATen/native/cuda/TensorCompare.cu [hunk 2]
```diff
@@ -43,7 +43,7 @@ void clamp_kernel_impl(TensorIteratorBase& iter) {
   AT_DISPATCH_ALL_TYPES_AND2(kHalf, kBFloat16, iter.common_dtype(), "clamp_cuda", [&] {
     gpu_kernel(iter, []GPU_LAMBDA(scalar_t v, scalar_t lower, scalar_t upper) -> scalar_t {
       // Propagate nan, which doesn't propagate automatically for ROCm
-      if (at::_isnan(v)) {
+      if (at::_isnan(v) || at::_isnan(lower) || at::_isnan(upper)) {
```
Comment (albanD) [inline aten/src/ATen/native/cuda/TensorCompare.cu:46]: The line below should be changed to a quiet nan?

**Review by albanD — COMMENTED** [review]
Sounds pretty good!

File: aten/src/ATen/native/cpu/TensorCompareKernel.cpp (hunk 1 above)
Comment (ngimel) [reply aten/src/ATen/native/cpu/TensorCompareKernel.cpp:345]: Yes

File: aten/src/ATen/native/cuda/TensorCompare.cu (hunk 2 above)
Comment (ngimel) [reply aten/src/ATen/native/cuda/TensorCompare.cu:46]: Oh you are right, except cuda doesn't like `std::numeric_limits<scalar_t>::quiet_NaN();` so it'll be a few ifs, will do.

**Review by albanD — APPROVED** [review]
SGTM !

**Pull request closed without merging**